platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -131,6 +152,212 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	uint8_t *iv_ptr;
+	uint64_t plaintext_bit_length = (param->cipher_range.length) * 8;
+	uint64_t iv_bit_length = (session->p.cipher_iv.length) * 8;
+	uint64_t aad_bit_length = (session->p.auth_aad_len) * 8;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+	int ret = 0;
+
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	cs.constants = &session->cc;
+
+	/* Set the counter */
+	if (armv8_aes_gcm_set_counter(iv_ptr, iv_bit_length, &cs) != 0) {
+		ODP_DBG("ARM Crypto: Failure while setting nonce\n");
+		ret = -1;
+		goto err;
+	}
+
+	cs.current_tag.d[0] = 0;
+	cs.current_tag.d[1] = 0;
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+	uint32_t len = in_len < seg_len ? in_len : seg_len;
+
+	if (in_len < seg_len) {
+		/* Packet is not segmented */
+		if (armv8_enc_aes_gcm_from_state(&cs,
+						 param->aad_ptr, aad_bit_length,
+						 data, plaintext_bit_length,
+						 local.output,
+						 local.digest) != 0) {
+			ODP_DBG("ARM Crypto: AES GCM Encoding failed\n");
+			ret = -1;
+			goto err;
+		}
+	} else {
+		/* Packet is segmented. Process the segments and copy the
+		 * contents to the input buffer */
+		memset(local.input, 0, sizeof(local.input));
+		uint8_t *cur_pos = local.input;
+
+		while (in_len > 0) {
+			len = in_len < seg_len ? in_len : seg_len;
+
+			odp_packet_copy_to_mem(pkt, in_pos, len, cur_pos);
+
+			/* Update the pointers */
+			in_len = in_len - len;
+			in_pos = in_pos + len;
+			cur_pos = cur_pos + len;
+		}
+
+		if (armv8_enc_aes_gcm_from_state(&cs,
```

Comment:
This library call does not need to be repeated in the two branches of the if statement. The only difference is the input data pointer that the else branch could just set to point to the local buffer.

Reply:
Sounds good; I didn't think of that. I will make the change.